src/lib/field.ts

```diff
@@ -0,0 +1,794 @@
+      this.value = x;
+      return;
+    }
+    // TODO this should handle common values efficiently by reading from a lookup table
+    this.value = FieldVar.constant(Fp(x));
+  }
+
+  // helpers
+  static #isField(x: bigint | number | string | Field | FieldVar): x is Field {
+    return x instanceof Field || (x as any) instanceof SnarkyFieldConstructor;
+  }
+  static #toConst(x: bigint | number | string | ConstantField): FieldConst {
+    if (Field.#isField(x)) return x.value[1];
+    return FieldConst.fromBigint(Fp(x));
+  }
+  static #toVar(x: bigint | number | string | Field): FieldVar {
+    if (Field.#isField(x)) return x.value;
+    return FieldVar.constant(Fp(x));
+  }
+  static from(x: bigint | number | string | Field): Field {
+    if (Field.#isField(x)) return x;
+    return new Field(x);
+  }
+
+  /**
+   * Checks whether this is a hard-coded constant in the Circuit.
+   */
+  isConstant(): this is { value: ConstantFieldVar } {
+    return this.value[0] === FieldType.Constant;
+  }
+
+  /**
+   * Returns a constant.
+   */
+  toConstant(): ConstantField {
+    if (this.isConstant()) return this;
+    // TODO: fix OCaml error message, `Can't evaluate prover code outside an as_prover block`
+    let value = Snarky.field.readVar(this.value);
+    return new Field(FieldVar.constant(value)) as ConstantField;
+  }
+
+  /**
+   * Serialize this instance of a {@link Field} to a bigint.
+   * This operation does NOT affect the circuit and can't be used to prove anything about the bigint representation of the Field.
+   */
+  toBigInt() {
+    let x = this.toConstant();
+    return FieldConst.toBigint(x.value[1]);
+  }
+
+  /**
+   * Serialize the {@link Field} to a string, e.g. for printing.
+   * This operation does NOT affect the circuit and can't be used to prove anything about the string representation of the Field.
+   */
+  toString() {
+    return this.toBigInt().toString();
+  }
+
+  /**
+   * Assert that this {@link Field} equals another Field-like value.
+   * Throws an error if the assertion fails.
+   *
+   * ```ts
+   * Field(1).assertEquals(1);
+   * ```
+   */
+  assertEquals(y: Field | bigint | number | string, message?: string) {
+    try {
+      if (this.isConstant() && isConstant(y)) {
+        if (this.toBigInt() !== toFp(y)) {
+          throw Error(`Field.assertEquals(): ${this} != ${y}`);
+        }
+        return;
+      }
+      Snarky.field.assertEqual(this.value, Field.#toVar(y));
+    } catch (err) {
+      throw withMessage(err, message);
+    }
+  }
+
+  /**
+   * Adds this {@link Field} element to another to a {@link Field} element.
+   *
+   * ```ts
+   * let a = Field(3);
+   * let sum = a.add(5)
+   * ```
+   */
+  add(y: Field | bigint | number | string): Field {
+    if (this.isConstant() && isConstant(y)) {
+      return new Field(Fp.add(this.toBigInt(), toFp(y)));
+    }
+    // return new AST node Add(x, y)
+    let z = Snarky.field.add(this.value, Field.#toVar(y));
+    return new Field(z);
+  }
```

Comment:
We could do that, but I don't feel it would make the code better. IMO, patterns are often better than abstractions, as in this case. 

Abstractions use specialized concepts (`binop(x, y, tsImpl, snarkyImpl)`) that have to be learned, while patterns use plain language features like if conditions that everyone already knows

Abstractions get leaky, or are only partially applicable. Patterns are flexible to adapt to any use case.

Unrelated reason why I'm not in favor is that I try to avoid all wrapper functions, because they add a line to our stack traces and our stack traces are too long.